#170947433 Implement user registration

AbdoulNuru opened the pull request:
#### What does this PR do?

- Implement user registration feature
- Use a 3rd API to send account verification emails.

#### How should this be manually tested?

- clone this repo
- install the dependencies by running `npm i` in your terminal.
- by referencing .env.example create a .env file and add all the keys.
- run `npm run migrateDb` in your terminal to run the migrations
- run `npm run dev` in your terminal to start the server.
- use postman to test this route `/api/v1/auth/signup` you will have to provide the following fields: firstName, lastName, email, password.
- check your email inbox, you will find an email from barefoot nomad.

#### Implementation plan

[User registration](https://docs.google.com/document/d/1qByH2hXJoWXIr9jFQIi3Co-gRLSjU_amzJQ3PnHsHDk/edit)

#### Stories associated with PR

[#170947433](https://www.pivotaltracker.com/story/show/170947433)

#### Screenshots if any
<img width="862" alt="Screen Shot 2020-02-11 at 13 31 22" src="https://user-images.githubusercontent.com/53453380/74233246-0130b080-4cd3-11ea-9843-0d7ca82928e5.png">
<img width="1108" alt="Screen Shot 2020-02-11 at 13 32 28" src="https://user-images.githubusercontent.com/53453380/74233260-042ba100-4cd3-11ea-861d-ddd4ad3d29eb.png">

Commit:
feature(signup): implement user registration

- modify the users table structure
- create sign up endpoint
- create a method to send verification emails
- create users route

[Maintains #170947433]

blaise82 commented on src/database/seeders/20200206095927-dummy-user.js:
```diff
@@ -1,8 +1,11 @@
+    password: '123',
+    isVerified: false,
     createdAt: new Date(),
     updatedAt: new Date()
```
is it possible that you may add a `method` too?

blaise82 replied:
is it possible that you may add a hashed password 
for **'password: '123''** won't have any use

AbdoulNuru replied:
@blaise82 Thank you, it is done

hakoemmy commented on src/tests/signup.test.js:
```diff
@@ -0,0 +1,41 @@
+Chai.use(chaiHttp);
+Chai.should();
+
+const user = {
```
@AbdoulNuru I think you can put this mock data in the mock folder
Put this 
```suggestion
const user = {
  firstName: 'Nuru',
  lastName: 'Abd',
  email: 'abdoulniyigena@gmail.com',
  password: 'nuru1234'
};
```
in file
   ``
tests/mockData/user.mock.js `` like that.

AbdoulNuru replied:
@hakoemmy Thank you, it is done

kcharles52 commented on package.json:
```diff
@@ -4,24 +4,28 @@
+    "start": "node --require @babel/polyfill build/index.js",
     "dev": "nodemon --exec babel-node src/index.js",
-    "test": "nyc --reporter=html --reporter=text mocha  --require @babel/register 'src/**/**test.js' --exit",
+    "test": "NODE_ENV=test npm run migrateTest && NODE_ENV=test nyc --reporter=html --reporter=text mocha  --require @babel/polyfill --require @babel/register 'src/**/**test.js' --exit",
```
```suggestion
    "test": "NODE_ENV=test npm run migrateTest && nyc --reporter=html --reporter=text mocha  --require @babel/polyfill --require @babel/register 'src/**/**test.js' --exit",
```

kcharles52 commented on src/controllers/user.controller.js:
```diff
@@ -0,0 +1,81 @@
+      const {
+        firstName, lastName, email, password
+      } = req.body;
+
```
```suggestion
```

kcharles52 commented on src/controllers/user.controller.js:
```diff
@@ -0,0 +1,81 @@
+      const newUser = await models.Users.create({
+        userID: uuid(), method: 'local', firstName, lastName, email, password, isVerified: false
+      });
+
```
```suggestion
```

remove these spaces

kcharles52 commented on src/routes/api/user.routes.js:
```diff
@@ -0,0 +1,57 @@
+ *             type: object
+ *             properties:
+ *               status:
+ *                 type: string
```
Please check the status type

blaise82 committed:
feat(user): social login with facebook and google (#13)

- create user routes
- create passport config
- create user controller

[Finishes #170947436]
